# Pandas Applied — 주문 데이터 분석

## 업무 요청

온라인 스토어의 한 달치 주문 데이터를 받았습니다.

> **이번 달 무엇이 매출을 만들었고, 이 데이터를 어디까지 믿어도 될까요?**

코드 실행에서 끝나지 않고 이 질문에 답할 Evidence를 만듭니다.

In [77]:
# colab 사용하시는 분들만!!!!
# 아래 코드 주석 해제해서 돌리셈

# import os
# os.makedirs('data', exist_ok=True)
#
# from google.colab import files
# uploaded = files.upload()   # 파일 선택창이 뜹니다
#
# # 업로드된 파일을 data/ 로 옮기기
# for name in uploaded:
#     os.rename(name, f'data/{name}')
#
# print(os.listdir('data'))

In [78]:
import pandas as pd

# Mission 1 — 데이터 읽기

`data/orders.csv`를 읽어 `orders`에 저장해보세요.

In [79]:
# CSV를 읽어 orders 변수에 저장하세요.
orders = pd.read_csv("data/orders.csv")

<details>
<summary>힌트 보기</summary>

`pd.read_csv("data/orders.csv")` 형태를 떠올려보세요.

</details>

## 바로 분석해도 될까요?

먼저 데이터의 처음 몇 행, 크기, 열, 결측 여부와 dtype을 확인하세요.

In [80]:
# 처음 몇 행을 확인하세요.
orders.head()

,order_id,customer_id,order_date,product_id,product_name,category,price,quantity,discount,payment_method,region,channel,record_status
0,ORD000001,C0291,2026-07-16,A004,토트백,패션,32000,5,0.00,카드,부산,Web,verified
1,ORD000002,C0254,2026-07-31,A003,캡,패션,25000,1,0.00,카드,서울,Web,verified
2,ORD000003,C0223,2026-07-05,S004,플래너,문구,12000,7,0.05,기타,서울,App,verified
3,ORD000004,C0074,2026-07-12,E002,마우스,전자기기,29000,3,0.05,카카오페이,경기,App,verified
4,ORD000005,C0049,2026-07-02,E005,USB 허브,전자기기,39000,5,0.05,네이버페이,경기,Web,verified


In [81]:
# 몇 행 몇 열인지 확인하세요.
orders.shape

(1000, 13)

In [82]:
# 열과 결측 여부, dtype을 한 번에 확인해보세요.
orders.info()

<class 'pandas.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   order_id        1000 non-null   str    
 1   customer_id     1000 non-null   str    
 2   order_date      1000 non-null   str    
 3   product_id      1000 non-null   str    
 4   product_name    1000 non-null   str    
 5   category        1000 non-null   str    
 6   price           1000 non-null   int64  
 7   quantity        1000 non-null   int64  
 8   discount        960 non-null    float64
 9   payment_method  1000 non-null   str    
 10  region          985 non-null    str    
 11  channel         1000 non-null   str    
 12  record_status   1000 non-null   str    
dtypes: float64(1), int64(2), str(10)
memory usage: 170.7 KB


In [83]:
orders.head()

,order_id,customer_id,order_date,product_id,product_name,category,price,quantity,discount,payment_method,region,channel,record_status
0,ORD000001,C0291,2026-07-16,A004,토트백,패션,32000,5,0.00,카드,부산,Web,verified
1,ORD000002,C0254,2026-07-31,A003,캡,패션,25000,1,0.00,카드,서울,Web,verified
2,ORD000003,C0223,2026-07-05,S004,플래너,문구,12000,7,0.05,기타,서울,App,verified
3,ORD000004,C0074,2026-07-12,E002,마우스,전자기기,29000,3,0.05,카카오페이,경기,App,verified
4,ORD000005,C0049,2026-07-02,E005,USB 허브,전자기기,39000,5,0.05,네이버페이,경기,Web,verified


In [84]:
orders.shape

(1000, 13)

In [85]:
orders.info()

<class 'pandas.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   order_id        1000 non-null   str    
 1   customer_id     1000 non-null   str    
 2   order_date      1000 non-null   str    
 3   product_id      1000 non-null   str    
 4   product_name    1000 non-null   str    
 5   category        1000 non-null   str    
 6   price           1000 non-null   int64  
 7   quantity        1000 non-null   int64  
 8   discount        960 non-null    float64
 9   payment_method  1000 non-null   str    
 10  region          985 non-null    str    
 11  channel         1000 non-null   str    
 12  record_status   1000 non-null   str    
dtypes: float64(1), int64(2), str(10)
memory usage: 170.7 KB


### Raw Data Evidence

- Raw 행 수: ______
- 열 수: ______
- 결측치가 있어 보이는 열: __________________
- 날짜처럼 보이는 열: __________________

## 어떤 열이 중요할까요?

팀장의 질문에 답하기 위해 먼저 필요하다고 생각하는 열을 적어보세요.

____________________________________________

In [86]:
orders.columns

Index(['order_id', 'customer_id', 'order_date', 'product_id', 'product_name',
       'category', 'price', 'quantity', 'discount', 'payment_method', 'region',
       'channel', 'record_status'],
      dtype='str')

In [87]:
# product_name, category, price, quantity 열만 선택하세요.
pass

# Mission 2 — 익숙한 조건부터

가격이 50,000원 이상인 주문을 찾습니다. 어떤 열을 어떤 값과 비교해야 하나요?

In [88]:
# 각 주문의 price가 50,000원 이상인지 True/False로 확인하세요.
pass

In [89]:
# 가격이 50,000원 이상인 주문만 선택하세요.
pass

`score >= 80`과 `price >= 50000`은 어떤 점에서 같은가요?

In [90]:
# 전자기기 카테고리 주문만 선택하세요.
pass

In [91]:
# 전자기기 주문 중 price가 100,000원 이상인 주문을 찾으세요.
pass

서울에서 발생한 전자기기 주문을 찾으려면 어떤 두 조건이 필요할까요?

In [92]:
# 두 조건을 조합해 주문을 찾으세요.
pass

<details>
<summary>힌트 보기</summary>

region 조건과 category 조건을 각각 만든 뒤 `&`로 연결합니다.

</details>

# Mission 3 — 매출 열 만들기

현재 `sales` 열은 없습니다. 가격과 수량으로 할인 전 주문금액을 만들 수 있을까요?

In [93]:
# price × quantity를 계산하세요. 아직 새 열에 저장하지 마세요.
pass

In [94]:
orders["sales"] = orders["price"] * orders["quantity"]
orders[["price", "quantity", "sales"]].head()

,price,quantity,sales
0,32000,5,160000
1,25000,1,25000
2,12000,7,84000
3,29000,3,87000
4,39000,5,195000


In [95]:
# sales가 가장 큰 주문 5개를 확인하세요.
orders.sort_values('sales', ascending=False).head(5)

,order_id,customer_id,order_date,product_id,product_name,category,price,quantity,discount,payment_method,region,channel,record_status,sales
680,ORD000681,C0285,2026-07-29,E006,이어폰,전자기기,129000,10,0.2,카드,대구,Web,verified,1290000
413,ORD000414,C0276,2026-07-12,E006,이어폰,전자기기,129000,6,NaN,네이버페이,서울,Web,verified,774000
872,ORD000873,C0288,2026-07-04,E003,모니터,전자기기,239000,3,NaN,카카오페이,부산,Web,verified,717000
292,ORD000293,C0293,2026-07-13,E003,모니터,전자기기,239000,3,NaN,계좌이체,경기,Web,verified,717000
261,ORD000262,C0299,2026-07-22,E003,모니터,전자기기,239000,3,NaN,카드,서울,Web,verified,717000


## 할인까지 적용하면?

할인 적용 후 금액의 계산 구조는 `sales × (1 - discount)`입니다. 결과가 모두 숫자로 채워질지 먼저 예상하세요.

In [96]:
preview_final_sales = orders["sales"] * (1 - orders["discount"])
preview_final_sales.head(12)

0     160000.0
1      25000.0
2      79800.0
3      82650.0
4     185250.0
5      35100.0
6      49000.0
7      93100.0
8      49000.0
9      10800.0
10     78000.0
11     30000.0
dtype: float64

### discount에 결측치가 있다!
- (1 - Nan)은 알 수가 없다. 
- 따라서 1 - Nan은 Nan이 된다. 
- 이 값을 원래 가격에 곱해도 그 결과는 Nan이 나온다. 

In [97]:
# 그러니 Nan값이 있는 지 확인해보자. 
preview_final_sales.isna().sum()

np.int64(40)

코드 문법은 맞지만 일부 계산 결과가 비어 있습니다. 원인을 데이터에서 찾아봅시다.

# Mission 4 — 이 데이터를 믿어도 될까?

In [98]:
# 각 열의 결측치 개수를 확인하세요.
pass

<details>
<summary>힌트 보기</summary>

`isna()`의 결과를 열별로 더해보세요.

</details>

In [99]:
orders.isna().sum()

order_id           0
customer_id        0
order_date         0
product_id         0
product_name       0
category           0
price              0
quantity           0
discount          40
payment_method     0
region            15
channel            0
record_status      0
sales              0
dtype: int64

### Missing Evidence

- discount 결측: ______
- region 결측: ______

## 같은 NaN, 다른 의미

`discount`가 비어 있으면 할인 적용 기록이 없는 주문입니다. 반면 `region`이 비어 있으면 지역 정보가 기록되지 않은 상태입니다.

두 값을 똑같이 처리해야 할까요? 각각의 처리와 이유를 적어보세요.

### discount

처리: ____________________  
이유: ____________________

### region

처리: ____________________  
이유: ____________________

In [100]:
# discount의 결측치를 업무 의미에 맞게 처리하세요.
pass

In [101]:
# 결측치 처리는 그 경우의 수가 무수히 많다. 
# 하지만 이 경우, 할인율은 보통 0.0~0.3 사이다. 따라서 0으로 맞추는 것도 고려할 수 있는 방안 중 하나다. 
orders["discount"] = orders["discount"].fillna(0)
orders["discount"].isna().sum()

np.int64(0)

## S47 — final_sales 만들기

이 단계에서는 각 행의 계산이 더 이상 결측 때문에 비지 않는지만 확인합니다. 아직 전체 공식 합계를 확정하지 않습니다.

In [102]:
# sales와 discount로 final_sales 열을 만드세요.
pass

In [103]:
orders["final_sales"] = orders["sales"] * (1 - orders["discount"])
orders["final_sales"].isna().sum()

np.int64(0)

## 날짜처럼 보이는 값은 정말 날짜일까요?

화면 모양과 실제 dtype은 다를 수 있습니다.

In [104]:
# order_date의 dtype을 확인하세요.
# dtype: str이다!!! 
a = orders["order_date"].head(3).tolist()
b = orders["order_date"]
print(a)
print(b)
print(type(a))
print(type(b))
print(orders['order_date'].dtypes)

['2026-07-16', '2026-07-31', '2026-07-05']
0      2026-07-16
1      2026-07-31
2      2026-07-05
3      2026-07-12
4      2026-07-02
          ...    
995    2026-07-14
996    2026-07-30
997    2026-07-22
998    2026-07-03
999    2026-07-06
Name: order_date, Length: 1000, dtype: str
<class 'list'>
<class 'pandas.Series'>
str


In [105]:
orders["order_date"] = pd.to_datetime(orders["order_date"])
# pd.to_datetime()을 사용하면 datetime64[us]로 타입이 변환된다. 
a = orders["order_date"].head(3).tolist()
b = orders["order_date"]
print(a)
print(b)
print(type(a))
print(type(b))
print(orders['order_date'].dtypes)

[Timestamp('2026-07-16 00:00:00'), Timestamp('2026-07-31 00:00:00'), Timestamp('2026-07-05 00:00:00')]
0     2026-07-16
1     2026-07-31
2     2026-07-05
3     2026-07-12
4     2026-07-02
         ...    
995   2026-07-14
996   2026-07-30
997   2026-07-22
998   2026-07-03
999   2026-07-06
Name: order_date, Length: 1000, dtype: datetime64[us]
<class 'list'>
<class 'pandas.Series'>
datetime64[us]


## 지역 값 살펴보기

지역별 개수를 세어 같은 지역이 서로 다른 값처럼 저장된 경우가 있는지 찾아보세요.

In [106]:
# NaN을 포함해 region별 값의 개수를 확인하세요.
pass

In [126]:
orders["region"].value_counts(dropna=False)

region
서울     349
경기     221
부산     152
대전     111
대구      80
광주      72
NaN     15
Name: count, dtype: int64

사람 눈에는 같은 지역인데 Pandas가 다르게 센 값이 있나요? 왜 그럴까요?

In [ ]:
# region의 앞뒤 공백을 제거하세요.
# 따라서 데이터셋을 받으면 문자열 값에 공백이 있는 지 확인해야 한다!
pass

<details>
<summary>힌트 보기</summary>

문자열의 앞뒤 공백은 `.str.strip()`으로 정리할 수 있습니다.

</details>

In [109]:
orders["region"] = orders["region"].str.strip()
dirty_variants = ["서울 ", " 경기 ", "부산 ", " 대전", "대구  ", " 광주"]
orders["region"].isin(dirty_variants).sum()

np.int64(0)

공백이 포함된 dirty variant가 0개가 되었는지만 확인했습니다. 지역별 공식 개수는 아직 확정하지 않습니다.

## 같은 주문이 두 번 저장되었다면?

먼저 모든 값이 완전히 같은 행이 있는지 확인합니다.

In [110]:
# 완전히 같은 행이 몇 개인지 확인하세요.
pass

In [111]:
orders.duplicated().sum()

np.int64(5)

In [112]:
# 중복된 행들을 직접 확인하세요.
pass

## Row Duplicate와 Key Duplicate

- Row Duplicate: 모든 값이 같은 행이 반복됨
- Key Duplicate: `order_id`는 같지만 다른 값이 서로 다를 수 있음

In [113]:
# order_id가 중복된 주문이 있는지 확인하세요.
pass

In [114]:
orders.loc[orders["order_id"].duplicated(keep=False)].sort_values("order_id").head(20)

,order_id,customer_id,order_date,product_id,product_name,category,price,quantity,discount,payment_method,region,channel,record_status,sales,final_sales
127,ORD000128,C0213,2026-07-11,E004,웹캠,전자기기,89000,5,0.20,카카오페이,서울,Web,verified,445000,356000.0
992,ORD000128,C0213,2026-07-11,E004,웹캠,전자기기,89000,5,0.20,카카오페이,서울,Web,verified,445000,356000.0
172,ORD000173,C0160,2026-07-22,F006,견과류,식품,13000,3,0.10,카드,서울,Web,verified,39000,35100.0
997,ORD000173,C0160,2026-07-22,F006,견과류,식품,13000,4,0.10,기타,서울,App,ingestion_error,52000,46800.0
256,ORD000257,C0140,2026-07-08,S001,노트,문구,4500,1,0.05,카드,경기,Web,verified,4500,4275.0
993,ORD000257,C0140,2026-07-08,S001,노트,문구,4500,1,0.05,카드,경기,Web,verified,4500,4275.0
440,ORD000441,C0260,2026-07-27,L006,머그컵,생활,14000,3,0.00,네이버페이,서울,Web,verified,42000,42000.0
994,ORD000441,C0260,2026-07-27,L006,머그컵,생활,14000,3,0.00,네이버페이,서울,Web,verified,42000,42000.0
513,ORD000514,C0088,2026-07-03,E002,마우스,전자기기,29000,3,0.00,카카오페이,대전,App,verified,87000,87000.0
998,ORD000514,C0088,2026-07-03,E002,마우스,전자기기,29000,4,0.00,기타,대전,Web,ingestion_error,116000,116000.0


충돌 레코드의 `record_status`를 확인하세요. 단순히 먼저 나온 행을 고르는 대신, 어떤 레코드를 신뢰할지 근거를 사용해야 합니다.

# Analysis-Ready Gate

이제 raw 1,000행에서 분석에 사용할 논리 주문으로 상태를 전환합니다.

`orders`는 원본에서 시작한 작업 데이터이고, `logical_orders`는 중복과 충돌 판단을 반영한 분석용 데이터입니다.

In [115]:
verified_orders = orders.loc[
    orders["record_status"].ne("ingestion_error")
].copy()

logical_orders = (
    verified_orders
    .drop_duplicates(subset="order_id", keep="first")
    .copy()
)

pd.Series({
    "logical_rows": len(logical_orders),
    "unique_order_id": logical_orders["order_id"].nunique(),
})

logical_rows       992
unique_order_id    992
dtype: int64

행 수와 고유 주문 ID 수가 모두 992인지 확인한 뒤에만 공식 분석을 시작합니다.

In [116]:
gross_total = logical_orders["sales"].sum()
discount_total = (logical_orders["sales"] - logical_orders["final_sales"]).sum()
final_total = logical_orders["final_sales"].sum()

pd.Series({
    "gross_sales": int(gross_total),
    "discount_amount": int(discount_total),
    "final_sales": int(final_total),
})

gross_sales        78000000
discount_amount     4000000
final_sales        74000000
dtype: int64

### 첫 공식 매출 Evidence

- Gross Sales: **78,000,000원**
- Discount Amount: **4,000,000원**
- Final Sales: **74,000,000원**

이 숫자들은 `logical_orders`가 992개의 고유 주문으로 확정된 뒤 처음 사용합니다.

# Cleaning Evidence

| 문제 | 발견 방법 | 처리 | 이유 |
|---|---|---|---|
| discount 결측 | | | |
| region 결측 | | | |
| region 공백 | | | |
| 완전 중복 | | | |
| order_id 충돌 | | | |
| order_date dtype | | | |

# Mission 5 — 무엇이 매출을 만들었을까?

전체 매출 하나만으로는 어떤 카테고리가 매출을 만들었는지 알 수 없습니다. 카테고리별로 나누어 비교합니다.

In [117]:
logical_orders["final_sales"].sum()

np.float64(74000000.0)

카테고리가 100개라면 각 카테고리를 따로 필터링해 합계를 100번 계산하는 방법은 불편합니다.

In [118]:
logical_orders.groupby("category")

`데이터 → category별로 나눔 → 각 그룹에서 계산 → 결과 비교`

카테고리별 합계 결과가 몇 개 나올지 예상해보세요.

In [141]:
print(logical_orders.groupby("category")["final_sales"].sum())

print(type(logical_orders.groupby("category")["final_sales"].sum()))

category
문구       4840000.0
생활      13300000.0
식품       6760000.0
전자기기    34800000.0
패션      14300000.0
Name: final_sales, dtype: float64
<class 'pandas.Series'>


`category`별로 묶고, `final_sales`를 선택하고, 각 그룹의 합을 계산했습니다.

In [140]:
# 카테고리별 매출을 높은 순서대로 정렬하세요.
logical_orders.groupby("category")["final_sales"].sum().sort_values() # Series에 sort_values()를 할 때는 정렬 기준 컬럼을 주지 않아도 된다. 

category
문구       4840000.0
식품       6760000.0
생활      13300000.0
패션      14300000.0
전자기기    34800000.0
Name: final_sales, dtype: float64

## '가장 많이 팔린 카테고리'는 무슨 뜻일까요?

1. 주문 건수가 가장 많은 카테고리
2. 총매출이 가장 큰 카테고리

두 답이 같은지 확인하세요.

In [121]:
# logical_orders에서 카테고리별 주문 건수를 구하세요.
pass

In [122]:
# logical_orders에서 카테고리별 총매출을 구하세요.
pass

In [123]:
category_summary = (
    logical_orders
    .groupby("category")["final_sales"]
    .agg(["count", "sum", "mean"])
)
category_summary

,count,sum,mean
category,,,
문구,220,4840000.0,22000.000000
생활,190,13300000.0,70000.000000
식품,260,6760000.0,26000.000000
전자기기,180,34800000.0,193333.333333
패션,142,14300000.0,100704.225352


### Category Evidence

주문 건수 1위: __________________  
총매출 1위: __________________  
평균 주문금액 1위: __________________

`count`, `sum`, `mean`은 각각 어떤 질문에 답하나요?

In [124]:
# 평균 주문금액이 가장 높은 카테고리를 찾으세요.
pass

# Applied Challenge — App과 Web 중 어떤 채널이 더 중요할까요?

바로 답하지 말고, 먼저 '중요하다'를 어떤 숫자로 판단할지 정하세요.

1. 중요하다의 기준: 총 매출
2. 분석 결과: 총 주문 수는 App, 총 구매액은 App, 주문 별 평균 금액은 Web이 더 높다. 
3. 뭐가 더 중요할 지는 이 분석을 기반으로 추가자료를 조사해 심층적으로 조사해야 한다. 

In [142]:
# logical_orders를 기준으로 채널별 주문 건수, 총매출, 평균 주문금액을 비교하세요.
logical_orders.columns

Index(['order_id', 'customer_id', 'order_date', 'product_id', 'product_name',
       'category', 'price', 'quantity', 'discount', 'payment_method', 'region',
       'channel', 'record_status', 'sales', 'final_sales'],
      dtype='str')

In [143]:
logical_orders.head(3)

,order_id,customer_id,order_date,product_id,product_name,category,price,quantity,discount,payment_method,region,channel,record_status,sales,final_sales
0,ORD000001,C0291,2026-07-16,A004,토트백,패션,32000,5,0.00,카드,부산,Web,verified,160000,160000.0
1,ORD000002,C0254,2026-07-31,A003,캡,패션,25000,1,0.00,카드,서울,Web,verified,25000,25000.0
2,ORD000003,C0223,2026-07-05,S004,플래너,문구,12000,7,0.05,기타,서울,App,verified,84000,79800.0


In [157]:
channel_summary = (
    logical_orders
    .groupby("channel")["final_sales"]
    .agg(["count", "sum", "mean"])
)
display(channel_summary)

for m in ["count", "sum", "mean"]:
    print(f"{m} 1st :{channel_summary[m].idxmax()}")

,count,sum,mean
channel,,,
App,560,35840000.0,64000.000000
Web,432,38160000.0,88333.333333


count 1st :App
sum 1st :Web
mean 1st :Web


<details>
<summary>힌트 보기</summary>

카테고리에서 사용한 `groupby + agg` 구조를 channel에 다시 적용할 수 있습니다.

```python

category_summary = (
    logical_orders
    .groupby("category")["final_sales"]
    .agg(["count", "sum", "mean"])
)
category_summary
```
</details>


## Channel Evidence

### 주문 건수가 많은 채널
____________________ / 근거: ____________________

### 총매출이 높은 채널
____________________ / 근거: ____________________

### 평균 주문금액이 높은 채널
____________________ / 근거: ____________________

### 내가 더 중요하다고 본 채널과 이유
________________________________________________

# 이번 Notebook에서 한 일

`업무 질문 → 데이터 확인 → 이상 발견 → 의미 판단 → 정리 → Analysis-Ready Gate → 그룹 비교 → Evidence`

다음에는 주문 파일에 없는 고객 등급을 다른 파일에서 연결합니다.